compiler/noirc_frontend/src/hir_def/types.rs

```diff
@@ -949,18 +988,26 @@ impl Type {
                     }
                 }
             }
-            Type::TypeVariable(binding, TypeVariableKind::Normal) => {
-                let borrow = binding.borrow();
+            Type::TypeVariable(self_var, TypeVariableKind::Normal) => {
+                let borrow = self_var.borrow();
                 match &*borrow {
-                    TypeBinding::Bound(typ) => typ.try_bind_to_polymorphic_int(var, bindings),
+                    TypeBinding::Bound(typ) => {
+                        typ.try_bind_to_polymorphic_int(var, bindings, only_integer)
+                    }
                     // Avoid infinitely recursive bindings
                     TypeBinding::Unbound(id) if *id == target_id => Ok(()),
                     TypeBinding::Unbound(new_target_id) => {
-                        // IntegerOrField is more specific than TypeVariable so we bind the type
-                        // variable to IntegerOrField instead.
-                        let clone =
-                            Type::TypeVariable(var.clone(), TypeVariableKind::IntegerOrField);
-                        bindings.insert(*new_target_id, (binding.clone(), clone));
+                        let clone_kind = if only_integer {
+                            // Integer is more specific than TypeVariable so we bind the type
+                            // variable to Integer instead.
+                            TypeVariableKind::Integer
+                        } else {
+                            // IntegerOrField is more specific than TypeVariable so we bind the type
+                            // variable to IntegerOrField instead.
+                            TypeVariableKind::IntegerOrField
+                        };
```

Comment:
```suggestion
                        // Bind to the most specific type variable kind
                        let clone_kind = if only_integer {
                            TypeVariableKind::Integer
                        } else {
                            TypeVariableKind::IntegerOrField
                        };
```
I think we can avoid the repetition here